lib/json/json.c

```diff
@@ -131,19 +131,19 @@ static void *lexer_string(struct lexer *lexer)
 			case 't':
 				continue;
 			case 'u':
-				if (!isxdigit(next(lexer))) {
+				if (!isxdigit((unsigned char)next(lexer))) {
```

Comment:
I do not see any sense in casting it into unsigned char. The isxdigit function takes int type argument and checks if it is between '0' and '9' or 'a'-'f' or 'A' - 'F'. All of this ranges have most significant bit cleared so even if char is signed it would not change the result.

Reply:
changed to casting to int in all cases.

Reply:
Why even bother to cast if you can simplicity cast from char to int without any warnings nor errors? Is there anything that forbids us from using such implicit cast?

Reply:
My guess is that isxdigit(), etc. are defined as macros. Then pedantic warning which @galak mentions in the commit (-Werror=char-subscripts) breaks all the fun. Solutions: a) don't use crazy pedantic warnings (e.g. newlib authors apparently don't use it and feel well); b) patch newlib (not each and every code which uses it!) to behave "well" (in some definition of well).

Reply:
D'oh, comment from ctype.h:

~~~
/* These macros are intentionally written in a manner that will trigger
   a gcc -Wall warning if the user mistakenly passes a 'char' instead
   of an int containing an 'unsigned char'.  Note that the sizeof will
   always be 1, which is what we want for mapping EOF to __ctype_ptr__[0];
   the use of a raw index inside the sizeof triggers the gcc warning if
   __c was of type char, and sizeof masks side effects of the extra __c.
   Meanwhile, the real index to __ctype_ptr__+1 must be cast to int,
   since isalpha(0x100000001LL) must equal isalpha(1), rather than being
   an out-of-bounds reference on a 64-bit machine.  */
~~~

So, it's all done on purpose. Hail the modern programmers. Note that casting a (signed) char to an int sin't correct (`char(-1)` will lead to `(int)-1`). The original cast to the unsigned char was (unless someone things that `(int)(unsigned char)` is "correcter" than just `(unsigned char)`).

Reply:
Great. Now it makes sense. I have never think that way. Something very above normal type checking. Seems like an overkill, but if they decided to do something like that in newlib, we have to adopt.